src/mappings/entities/mapAsset.ts

```diff
@@ -0,0 +1,521 @@
+  await asset.save();
+};
+
+const handleMakeDivisible = async (params: Record<string, any>) => {
+  const { ticker } = params;
+  const asset = await getAsset(ticker);
+  asset.isDivisible = true;
+  await asset.save();
+};
+
+const handleIssue = async (params: Record<string, any>) => {
+  const { ticker, amount } = params;
+  const asset = await getAsset(ticker);
+  const formattedAmount = chainAmountToBigNumber(amount);
+  const ownerAmount = getHolderAmount(asset.ownerDid, asset.holders);
+  const otherHolders = asset.holders.filter((h) => h.did !== asset.ownerDid);
+  asset.holders = [
+    ...otherHolders,
+    {
+      did: asset.ownerDid,
+      amount: ownerAmount.plus(formattedAmount).toString(),
+    },
+  ];
+  asset.totalSupply = new BigNumber(asset.totalSupply)
+    .plus(formattedAmount)
+    .toString();
+  await asset.save();
+};
+
+const handleRedeem = async (params: Record<string, any>) => {
+  const { ticker, value: amount } = params;
+  const asset = await getAsset(ticker);
+  const formattedAmount = chainAmountToBigNumber(amount);
+  const ownerAmount = getHolderAmount(asset.ownerDid, asset.holders);
+  const otherHolders = asset.holders.filter((h) => h.did !== asset.ownerDid);
+  asset.holders = [
+    ...otherHolders,
+    {
+      did: asset.ownerDid,
+      amount: ownerAmount.minus(formattedAmount).toString(),
+    },
+  ];
+  asset.totalSupply = new BigNumber(asset.totalSupply)
+    .minus(formattedAmount)
+    .toString();
+  await asset.save();
+};
+
+const handleFreeze = async (params: Record<string, any>) => {
+  const { ticker } = params;
+  const asset = await getAsset(ticker);
+  asset.isFrozen = true;
+  await asset.save();
+};
+
+const handleUnfreeze = async (params: Record<string, any>) => {
+  const { ticker } = params;
+  const asset = await getAsset(ticker);
+  asset.isFrozen = false;
+  await asset.save();
+};
+
+const handleAcceptAssetOwnershipTransfer = async (
+  params: Record<string, any>,
+) => {
+  const { authId } = params;
+  const authorization = await getAuthorization(authId);
+  if (!authorization.ticker) return;
+  const asset = await getAsset(authorization.ticker);
+  asset.ownerDid = authorization.target;
+  await asset.save();
+  await Authorization.remove(authId);
+};
+// #endregion
+
+// #region ModuleIdEnum.Settlement
+const handleAddAndAffirmInstruction = async (
+  params: Record<string, any>,
+  extrinsic: any,
+) => {
+  const { legs } = params;
+  await Promise.all(
+    legs.map(async (l: any, i: number) => {
+      const settlementId = Number(extrinsic.events[0].event.data[2].toString());
+      await Settlement.create({
+        id: `${settlementId}-${i + 1}`,
+        settlementId,
+        from: l.from.did.toString(),
+        to: l.to.did.toString(),
+        ticker: l.asset.toString(),
+        amount: chainAmountToBigNumber(l.amount).toString(),
+      }).save();
+    }),
+  );
+};
+
+const handleAffirmInstruction = async (params: Record<string, any>) => {
```

Comment:
Each participant will affirm an instruction.

I think we want to wait for a completed instruction and not the affirmation for this.

Reply:
In this case it's affirming an instruction that was affirmed when added (`addAndAffirmInstruction`). This works with the UI but yeah would be better to also handle cases where someone might do `addInstruction` from Polkadot and affirm it later.

Reply:
even with `addAndAffirmInstruction`, the other party still needs to affirm it, so unless its an instruction thats acting a portfolio transfer wouldn't we run into the multi affirm issue?

Reply:
Yeah and this function (`handleAffirmInstruction`) handles the other party's affirmation. The initial party already affirmed it when they added it. We're only listening to `addAndAffirmInstruction` right now, and handling it using a separate function (`handleAddAndAffirmInstruction`). So this will handle all settlements that were added via `addAndAffirmInstruction`, which is what the UI does.

Reply:
What about instructions with multiple parties involved? There can be arbitrary parties involved in an instruction, I don't think we can assume its always 2 people. There can be a complex one with 5 parties involved and we'd get 4 affirmation in addition to the AddAndAffirm.

Reply:
Each party would be a different leg, right? How does it work on the chain side? Because you get one instruction ID for the legs. So when you affirm an instruction by providing the ID, does the chain wait for all parties to affirm the ID?
It handles multiple legs, but as soon as someone affirms it, it will process it.

Reply:
Apparently the chain will process each leg as it is affirmed. So will implement the change to process the leg for user who called affirm.

Reply:
There is one ID for everything, legs don't get an ID. My understanding is when you affirm an instruction the chain only marks that and only triggers the settlement if that was the last pending affirmation.

I looks like there is an  `InstructionExecuted` event, that we can use. When that fires all the legs have been executed atomically.

I think we also need `InstructionFailed`, `InstructionRejected` + `InstructionRescheduled` (Not sure if this would trigger a Created event as well though) events to keep track of Settlements that are complete as well.

Reply:
An affirmation causes is a very minimal state change unless it was the last pending one.

Then all the legs get executed and the assets get moved only on the Executed event.

Reply:
The point of having legs is to be able to facilitate complex transactions atomically. If they were executed one by one there wouldn't be much use for them imo.

Reply:
Refactored to use existing settlements on event side.